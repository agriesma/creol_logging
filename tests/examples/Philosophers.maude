load creol-interpreter .
mod PROGRAM is
  protecting CREOL-SIMULATOR .
  op classes : -> Configuration [ctor] .
  eq classes =
    < "Philosopher" : Class | Inh: noInh, Param: "butler", Att: "butler" |->
      null, "hungry" |-> null, "chopstick" |-> null, "neighbor" |-> null,
      "hist" |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "label:0" |-> null, "_"
              |-> null, Code: assign( "hist" ; str("") ) ;
              assign( "chopstick" ; bool(true) ) ;
              assign( "hungry" ; bool(false) ) ; call( "label:0" ; "butler" ;
              "getNeighbor" ; emp ) ; get( "label:0" ; "neighbor" ) ;
              free( "label:0" ) ; return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "label:2" |-> null,
              "label:1" |-> null, "label:0" |-> null, "_" |-> null,
              Code: call( "label:0" ; "this" ; "think" ; emp ) ;
              free( "label:0" ) ; call( "label:1" ; "this" ; "eat" ; emp ) ;
              free( "label:1" ) ; call( "label:2" ; "this" ; "digest" ; emp )
              ; free( "label:2" ) ; return ( emp ) >,
            < "digest" : Method | Param: noVid, Att: "label:0" |-> null, "_"
              |-> null, Code: release ; assign( "hungry" ; bool(true) ) ;
              assign( "hist" ; "|-" ( "hist" :: str("d") ) ) ; release ;
              call( "label:0" ; "this" ; "digest" ; emp ) ; free( "label:0" )
              ; return ( emp ) >,
            < "eat" : Method | Param: noVid, Att: "label:1" |-> null,
              "label:0" |-> null, "l" |-> null, "_" |-> null, Code: await
              "hungry" ; call( "l" ; "neighbor" ; "borrowStick" ; emp ) ;
              await ?("l") ; free( "l" ) ; await "chopstick" ;
              assign( "hist" ; "|-" ( "hist" :: str("e") ) ) ;
              assign( "hungry" ; bool(false) ) ; call( "label:0" ;
              "neighbor" ; "returnStick" ; emp ) ; free( "label:0" ) ;
              release ; call( "label:1" ; "this" ; "eat" ; emp ) ;
              free( "label:1" ) ; return ( emp ) >,
            < "think" : Method | Param: noVid, Att: "label:0" |-> null, "_"
              |-> null, Code: await "~" ( "hungry" ) ;
              assign( "hist" ; "|-" ( "hist" :: str("t") ) ) ; release ;
              call( "label:0" ; "this" ; "think" ; emp ) ; free( "label:0" )
              ; return ( emp ) >,
            < "borrowStick" : Method | Param: noVid, Att: "_" |-> null,
              Code: await "chopstick" ; assign( "chopstick" ; bool(false) ) ;
              return ( emp ) >,
            < "returnStick" : Method | Param: noVid, Att: "_" |-> null,
              Code: assign( "chopstick" ; bool(true) ) ; return ( emp ) >,
      Ocnt: 0 >
    
    < "Butler" : Class | Inh: noInh, Param: noVid, Att: "p1" |-> null, "p2"
      |-> null, "p3" |-> null, "p4" |-> null, "p5" |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "_" |-> null,
              Code: new( "p1" ; "Philosopher" ; "this" ) ;
              new( "p2" ; "Philosopher" ; "this" ) ;
              new( "p3" ; "Philosopher" ; "this" ) ;
              new( "p4" ; "Philosopher" ; "this" ) ;
              new( "p5" ; "Philosopher" ; "this" ) ; return ( emp ) >,
            < "getNeighbor" : Method | Param: noVid, Att: "neighbor" |->
              null, "_" |-> null, Code: if "=" ( "caller" :: "p1" ) th
              assign( "neighbor" ; "p2" ) el if "=" ( "caller" :: "p2" ) th
              assign( "neighbor" ; "p3" ) el if "=" ( "caller" :: "p3" ) th
              assign( "neighbor" ; "p4" ) el if "=" ( "caller" :: "p4" ) th
              assign( "neighbor" ; "p5" ) el assign( "neighbor" ; "p1" ) fi
              fi fi fi ; return ( "neighbor" ) >,
      Ocnt: 0 > .
endm

