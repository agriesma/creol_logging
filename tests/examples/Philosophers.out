==========================================
reduce in PROGRAM : classes .
result Configuration: 
< "Butler" : Class | Inh: noInh, Param: noVid, Att: "p1" |-> null, "p2" |->
    null, "p3" |-> null, "p4" |-> null, "p5" |-> null, Mtds:
  < "getNeighbor" : Method | Param: noVid, Att: "_" |-> null, "neighbor" |->
    null, Code: if "="("caller" :: "p1") th assign("neighbor" ; "p2") el if
    "="("caller" :: "p2") th assign("neighbor" ; "p3") el if "="("caller" ::
    "p3") th assign("neighbor" ; "p4") el if "="("caller" :: "p4") th assign(
    "neighbor" ; "p5") el assign("neighbor" ; "p1") fi fi fi fi ; return(
    "neighbor") >, < "init" : Method | Param: noVid, Att: "_" |-> null, Code:
    skip ; return(emp) >, < "run" : Method | Param: noVid, Att: "_" |-> null,
    Code: new("p1" ; "Philosopher" ; "this") ; new("p2" ; "Philosopher" ;
    "this") ; new("p3" ; "Philosopher" ; "this") ; new("p4" ; "Philosopher" ;
    "this") ; new("p5" ; "Philosopher" ; "this") ; return(emp) >,
  Ocnt: 0 >
 
< "Philosopher" : Class | Inh: noInh, Param: "butler", Att: "butler" |-> null,
    "chopstick" |-> null, "hist" |-> null, "hungry" |-> null, "neighbor" |->
    null, Mtds:
  < "borrowStick" : Method | Param: noVid, Att: "_" |-> null, Code: await
    "chopstick" ; assign("chopstick" ; bool(false)) ; return(emp) >, < "digest"
    : Method | Param: noVid, Att: "_" |-> null, "label:0" |-> null, Code:
    release ; assign("hungry" ; bool(true)) ; assign("hist" ; "|-"("hist" ::
    str("d"))) ; release ; call("label:0" ; "this" ; "digest" ; emp) ; free(
    "label:0") ; return(emp) >, < "eat" : Method | Param: noVid, Att: "_" |->
    null, "l" |-> null, "label:0" |-> null, "label:1" |-> null, Code: await
    "hungry" ; call("l" ; "neighbor" ; "borrowStick" ; emp) ; await ?("l") ;
    free("l") ; await "chopstick" ; assign("hist" ; "|-"("hist" :: str("e"))) ;
    assign("hungry" ; bool(false)) ; call("label:0" ; "neighbor" ;
    "returnStick" ; emp) ; free("label:0") ; release ; call("label:1" ; "this"
    ; "eat" ; emp) ; free("label:1") ; return(emp) >, < "init" : Method |
    Param: noVid, Att: "_" |-> null, "label:0" |-> null, Code: assign("hist" ;
    str("")) ; assign("chopstick" ; bool(true)) ; assign("hungry" ; bool(
    false)) ; call("label:0" ; "butler" ; "getNeighbor" ; emp) ; get("label:0"
    ; "neighbor") ; free("label:0") ; return(emp) >, < "returnStick" : Method |
    Param: noVid, Att: "_" |-> null, Code: assign("chopstick" ; bool(true)) ;
    return(emp) >, < "run" : Method | Param: noVid, Att: "_" |-> null,
    "label:0" |-> null, "label:1" |-> null, "label:2" |-> null, Code: call(
    "label:0" ; "this" ; "think" ; emp) ; free("label:0") ; call("label:1" ;
    "this" ; "eat" ; emp) ; free("label:1") ; call("label:2" ; "this" ;
    "digest" ; emp) ; free("label:2") ; return(emp) >, < "think" : Method |
    Param: noVid, Att: "_" |-> null, "label:0" |-> null, Code: await "~"(
    "hungry") ; assign("hist" ; "|-"("hist" :: str("t"))) ; release ; call(
    "label:0" ; "this" ; "think" ; emp) ; free("label:0") ; return(emp) >,
  Ocnt: 0 >

Bye.
